src/python/pants/engine/internals/selectors_test.py

```diff
@@ -1,12 +1,77 @@
 # Copyright 2016 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+import ast
 import re
-from typing import Any
+from typing import Any, Tuple
 
 import pytest
 
-from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.internals.selectors import Get, GetConstraints, GetParseError, MultiGet
+
+
+def parse_get_types(get: str) -> Tuple[str, str]:
+    get_args = ast.parse(get).body[0].value.args  # type: ignore[attr-defined]
+    return GetConstraints.parse_product_and_subject_types(get_args, source_file_name="test.py")
+
+
+def test_parse_get_types_valid() -> None:
+    assert parse_get_types("Get(P, S, subject)") == ("P", "S")
+    assert parse_get_types("Get(P, S())") == ("P", "S")
+
+
+def assert_parse_get_types_fails(get: str, *, expected_explanation: str) -> None:
+    with pytest.raises(GetParseError) as exc:
+        parse_get_types(get)
+    assert str(exc.value) == f"Invalid Get. {expected_explanation} Failed for {get} in test.py."
+
+
+def test_parse_get_types_wrong_number_args() -> None:
+    assert_parse_get_types_fails(
+        "Get()",
+        expected_explanation="Expected either two or three arguments, but got 0 arguments.",
```

Comment:
I know this is brittle to be doing exact string matches. But because we're trying to improve the quality of our error messages, I think this fragility is justified so that we are confident the error message is rendering exactly how we'd like.